scvi/models/modules.py

```diff
@@ -8,88 +8,51 @@
 
 
 class FCLayers(nn.Module):
-    def __init__(self, n_in, n_out, n_hidden=128, n_layers=1, dropout_rate=0.1):
+    def __init__(self, n_in, n_out, n_cat_list=[], n_layers=1, n_hidden=128, dropout_rate=0.1):
         super(FCLayers, self).__init__()
         layers_dim = [n_in] + (n_layers - 1) * [n_hidden] + [n_out]
-        self.fc_layers = FCLayers._sequential(layers_dim, dropout_rate=dropout_rate)
-
-    def forward(self, x, *os):
-        return self.fc_layers(x)
-
-    @staticmethod
-    def create(n_in, n_out, n_cat=0, n_hidden=128, n_layers=1, dropout_rate=0.1):
-        if type(n_cat) is int:
-            if n_cat == 0:
-                return FCLayers(n_in, n_out, n_hidden=n_hidden, n_layers=n_layers, dropout_rate=dropout_rate)
-            else:
-                return OneHotFCLayers(n_in, n_out, n_cat=n_cat, n_hidden=n_hidden, n_layers=n_layers,
-                                      dropout_rate=dropout_rate)
-        elif type(n_cat) is list:
-            return ManyOneHotFCLayers(n_in, n_out, n_cat_list=n_cat,
-                                      n_hidden=n_hidden, n_layers=n_layers, dropout_rate=dropout_rate)
-
-    @staticmethod
-    def _sequential(layers_dim, n_cat=0, dropout_rate=0.1):
-        return nn.Sequential(collections.OrderedDict(
+        self.n_cat_list = [n_cat if n_cat > 1 else 0 for n_cat in n_cat_list]  # n_cat = 1 will be ignored
+        self.fc_layers = nn.Sequential(collections.OrderedDict(
             [('Layer {}'.format(i), nn.Sequential(
                 nn.Dropout(p=dropout_rate),
-                nn.Linear(n_in + n_cat, n_out),
+                nn.Linear(n_in + sum(n_cat_list), n_out),
                 nn.BatchNorm1d(n_out, eps=1e-3, momentum=0.99),
```

Comment:
Dropout usually happens after fully connected layers--not before. Should these be reordered?

Reply:
This was the `dense` layer implementation in Romain's code. 

> Typically, dropout is applied after the non-linear activation function (a). However, when using rectified linear units (ReLUs), it might make sense to apply dropout before the non-linear activation (b) for reasons of computational efficiency depending on the particular code implementation.

in https://sebastianraschka.com/faq/docs/dropout-activation.html.

Reply:
We could move it btw Linear and BatchNorm, or after BatchNorm ? What would you suggest, and what results should we look at for making such change ? I could always run the mixing experiment which results seem to be dropout-dependent. (It always messes up for dropout=0)